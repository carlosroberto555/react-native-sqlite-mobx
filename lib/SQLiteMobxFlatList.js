import React, { useEffect } from 'react';
import { FlatList } from 'react-native';
import { Observer } from 'mobx-react-lite';
function SQLiteMobxFlatList(props) {
    const { model, where, ...rest } = props;
    useEffect(() => {
        model.where = where;
        model.loadItems();
    }, [where]);
    return (<Observer>
			{() => (<FlatList {...rest} data={model.data} initialNumToRender={model.data.length}/>)}
		</Observer>);
}
export default SQLiteMobxFlatList;
//# sourceMappingURL=SQLiteMobxFlatList.js.map