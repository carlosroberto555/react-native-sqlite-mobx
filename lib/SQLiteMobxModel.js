var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { action, runInAction } from 'mobx';
import SQLite from './SQLite';
export default class SQLiteMobxModel {
    async loadItems() {
        const resp = await SQLite.query(`SELECT * FROM ${this.table} WHERE ${this.where || 1}`);
        runInAction(() => {
            this.data = resp.toArray() || [];
        });
    }
    async setItems(items) {
        await SQLite.truncateTable(this.table);
        await SQLite.insertMany(this.table, items);
        await this.loadItems();
    }
    async addItems(items) {
        await SQLite.insertMany(this.table, items);
        await this.loadItems();
    }
    async addItem(item) {
        await SQLite.insert(this.table, item);
        await this.loadItems();
    }
    async getItem(id) {
        // TODO: implement this
    }
    async editItem(item) {
        // TODO: implement this
    }
    async removeItem(item) {
        await SQLite.query(`DELETE FROM ${this.table} WHERE id = ${item.id}`);
        await this.loadItems();
    }
}
__decorate([
    action
], SQLiteMobxModel.prototype, "loadItems", null);
__decorate([
    action
], SQLiteMobxModel.prototype, "setItems", null);
__decorate([
    action
], SQLiteMobxModel.prototype, "addItems", null);
__decorate([
    action
], SQLiteMobxModel.prototype, "addItem", null);
__decorate([
    action
], SQLiteMobxModel.prototype, "getItem", null);
__decorate([
    action
], SQLiteMobxModel.prototype, "editItem", null);
__decorate([
    action
], SQLiteMobxModel.prototype, "removeItem", null);
//# sourceMappingURL=SQLiteMobxModel.js.map