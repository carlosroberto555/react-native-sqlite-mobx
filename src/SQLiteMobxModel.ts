import { action, runInAction, observable } from 'mobx'
import SQLite from './SQLite'

type QueryParams = {
	select?: string
	join?: string
	where?: string
}

interface SQLiteItem {
	[key: string]: number | string
	id: number
}

export default abstract class SQLiteMobxModel<T extends object> {
	@observable lastParams?: QueryParams
	abstract table: string
	abstract data: (T & SQLiteItem)[]

	@action.bound
	async loadItems(params?: QueryParams) {
		const { select, join, where } = getQueryParams(params)

		const resp = await SQLite.query(
			`SELECT ${select} FROM ${this.table} ${join} WHERE ${where}`
		)
		runInAction(() => {
			this.data = resp.toArray() || []
			this.lastParams = params
		})
	}

	@action.bound
	async setItems(items: T[]) {
		await SQLite.truncateTable(this.table)
		await SQLite.insertMany(this.table, items)
		this.loadItems(this.lastParams)
	}

	@action.bound
	async addItems(items: T[]) {
		await SQLite.insertMany(this.table, items)
		this.loadItems(this.lastParams)
	}

	@action.bound
	async addItem(item: T) {
		await SQLite.insert(this.table, item)
		this.loadItems(this.lastParams)
	}

	@action.bound
	async setItem(item: T & SQLiteItem) {
		await SQLite.insertOrReplace(this.table, item)
		this.loadItems(this.lastParams)
	}

	@action.bound
	async getItem(id: number) {
		// TODO: implement this
	}

	@action.bound
	async editItem(item: T) {
		// TODO: implement this
	}

	@action.bound
	async removeItem(item: T & SQLiteItem) {
		await SQLite.query(`DELETE FROM ${this.table} WHERE id = ${item.id}`)
		this.loadItems(this.lastParams)
	}

	@action.bound
	clear() {
		this.lastParams = undefined
		this.data = []
	}
}

function getQueryParams(p?: QueryParams) {
	if (p) {
		return {
			select: p.select || '*',
			join: p.join || '',
			where: p.where || '1'
		}
	} else {
		return {
			select: '*',
			join: '',
			where: '1'
		}
	}
}
