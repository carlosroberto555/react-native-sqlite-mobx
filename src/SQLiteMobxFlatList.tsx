import React, { useEffect } from 'react'
import { FlatList, FlatListProps } from 'react-native'
import { useObserver } from 'mobx-react-lite'

import SQLiteMobxModel from './SQLiteMobxModel'

interface SQLiteMobxFlatListProps<T extends { id: number }>
	extends Omit<FlatListProps<T>, 'data'> {
	model: SQLiteMobxModel<T>
	autofetch?: boolean
	select?: string
	where?: string
	join?: string
}

function SQLiteMobxFlatList(props: SQLiteMobxFlatListProps<any>) {
	const { model, select, where, join, autofetch = true, ...rest } = props

	if (autofetch) {
		useEffect(() => {
			model.loadItems({ select, join, where })
			return () => model.clear()
		}, [select, join, where])
	}

	return useObserver(() => <FlatList {...rest} data={model.data} />)
}

export { SQLiteMobxFlatListProps }
export default SQLiteMobxFlatList
