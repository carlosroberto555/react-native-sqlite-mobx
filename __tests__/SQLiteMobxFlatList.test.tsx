import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'

import SQLiteMobxFlatList from '../src/SQLiteMobxFlatList'
import SQLiteMobxModel from '../src/SQLiteMobxModel'

type ModelDataType = {
	id: number
}

class ModelTest extends SQLiteMobxModel<ModelDataType> {
	data: ModelDataType[] = []
	table: string = 'test'
}

const modelTest = new ModelTest()

test('SQLiteMobxFlatList renders correctly', () => {
	const tree = renderer.create(
		<SQLiteMobxFlatList model={modelTest} renderItem={() => null} />
	)

	expect(tree).toMatchSnapshot()
})
