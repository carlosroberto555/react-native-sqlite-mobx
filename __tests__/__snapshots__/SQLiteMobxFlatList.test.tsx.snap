// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SQLiteMobxFlatList renders correctly 1`] = `
<RCTScrollView
  data={Array []}
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View />
</RCTScrollView>
`;
